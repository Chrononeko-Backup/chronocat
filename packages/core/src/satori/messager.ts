import Element from '@satorijs/element'
import type { Object } from 'ts-toolbelt'
import type { Common } from '../common'
import type { ChronocatSatoriServerConfig } from '../config/types'
import { MessageRecvDispatchMessage } from '../dispatch'
import type { Peer, Element as RedElement } from '../red'
import { ChatType, r } from '../red'
import type { Message } from './types'

class State {
  constructor(public type: 'message') {}
}

export class Messager {
  constructor(
    public selfId: string,
    public config: ChronocatSatoriServerConfig,
    public common: Common,
    public channelId: string,
  ) {
    this.peer = this.channelId.startsWith('private:')
      ? {
          chatType: ChatType.Private,
          peerUin: this.channelId.slice(8), // private:
        }
      : {
          chatType: ChatType.Group,
          peerUin: this.channelId,
          guildId: '',
        }
  }

  peer: Partial<Peer>

  public errors: Error[] = []
  public results: Message[] = []

  stack: State[] = [new State('message')]
  children: Object.Partial<RedElement, 'deep'>[] = []

  prepare = async () => {}

  render = async (elements: Element[], flush?: boolean) => {
    for (const element of elements) await this.visit(element)
    if (flush) await this.flush()
  }

  async send(content: Element.Fragment) {
    await this.prepare()
    const elements = Element.normalize(content)
    await this.render(elements)
    await this.flush()

    if (this.errors.length) throw this.errors
    else return this.results.filter(Boolean) // .map(({ id }) => ({ id }))
  }

  flush = async () => {
    if (!this.children.length) return

    const result = await this.common.send(this.peer, this.children)
    const parsedEvents = await new MessageRecvDispatchMessage([
      result,
    ]).toSatori(this.selfId, this.config)
    for (const parsedEvent of parsedEvents)
      if (parsedEvent.message) this.results.push(parsedEvent.message)
    this.children = []
  }

  visit = async (element: Element) => {
    const { type, attrs, children } = element

    switch (type) {
      case 'text': {
        // 文本消息
        this.children.push(r.text(attrs['content'] as string))
        return
      }

      case 'img': {
        // 图片消息
        const urlString = attrs['src'] as string
        const result = await this.common.save(urlString)

        let picType: number
        switch (result.imageInfo?.type) {
          case 'png': {
            picType = 1001
            return
          }

          case 'gif': {
            picType = 2000
            return
          }

          default: {
            picType = 1000
            break
          }
        }

        this.children.push(r.remoteImage(result, picType))
        return
      }

      case 'audio': {
        // 语音消息
        const urlString = attrs['src'] as string
        const result = await this.common.save(urlString)
        this.children.push(
          r.remoteAudio(
            result,
            (attrs['chrono-unsafe-time'] as number | undefined) || 1,
          ),
        )
        return
      }

      case 'file': {
        // 文件消息
        const urlString = attrs['src'] as string
        const result = await this.common.save(urlString)
        this.children.push(r.remoteFile(result))
        return
      }

      case 'at': {
        // at 消息
        if (attrs['type'] === 'all') {
          this.children.push(r.at('全体成员', 'all'))
        } else {
          this.children.push(
            r.at(attrs['name'] as string, attrs['id'] as string),
          )
        }

        return
      }

      case 'message': {
        // 前面的消息直接发送，开始一条新消息
        await this.flush()

        if ('forward' in attrs) {
          // TODO: 合并转发
          // this.stack.unshift(new State('forward'))
          // await this.render(children)
          // await this.flush()
          // this.stack.shift()
          // await this.forward()
        } else {
          // 普通切割消息
          await this.render(children, true)
        }
        return
      }

      default: {
        // 兜底
        await this.render(children)
        return
      }
    }
  }
}
