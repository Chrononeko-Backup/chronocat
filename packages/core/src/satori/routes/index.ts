import { guildMemberList } from './guild/member/list'
import { loginGet } from './login/get'
import { messageCreate } from './message/create'
import type { Route } from './types'
import { userChannelCreate } from './user/channel/create'

const routesIntl = {
  'guild.member.list': guildMemberList,
  'login.get': loginGet,
  'message.create': messageCreate,
  'user.channel.create': userChannelCreate,
} as const

export type Routes = keyof typeof routesIntl

export const routes: Record<Routes, Route> = routesIntl
