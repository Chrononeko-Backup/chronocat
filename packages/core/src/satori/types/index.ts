import type Element from '@satorijs/element'
import type { GuildMember } from './satori'

export * from './satori'

export type WebSocketIncomingMessage =
  | WebSocketIncomingHeartbeatMessage
  | WebSocketIncomingVerifyMessage

export enum Op {
  Event = 0,
  Ping = 1,
  Pong = 2,
  Identify = 3,
  Ready = 4,
}

export interface WebSocketIncomingHeartbeatMessage {
  op: Op.Ping
  body: never
}

export interface WebSocketIncomingVerifyMessage {
  op: Op.Identify
  body?: {
    token?: string
  }
}

export interface Next {
  next?: string
}

export interface MessageCreatePayload {
  channel_id: string
  content: Element.Fragment
}

export interface UserChannelCreatePayload {
  user_id: string
}

export interface GuildMemberListPayload extends Next {
  guild_id: string
}

export interface GuildMemberListResponse extends Next {
  data: GuildMember[]
}
