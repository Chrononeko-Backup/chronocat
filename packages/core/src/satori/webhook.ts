import fetch from 'node-fetch'
import type { ChronocatSatoriWebHookConfig } from '../config/types'
import type { DispatchMessage } from '../dispatch'
import { getAuthData } from '../utils/authData'
import { buildEventIdCounter } from '../utils/token'

export const initSatoriWebHook = async (
  config: ChronocatSatoriWebHookConfig,
) => {
  // 预处理 self_url
  if (!config.self_url || config.self_url === 'https://chronocat.vercel.app')
    config.self_url = `http://127.0.0.1:5500`
  if (config.self_url.endsWith('/'))
    config.self_url = config.self_url.slice(0, config.self_url.length - 1)

  const getId = buildEventIdCounter()

  const dispatcher = async (message: DispatchMessage) => {
    await message.toSatori((await getAuthData()).uin, config).then((events) =>
      events.forEach((data) => {
        const body = {
          ...data,
          id: getId(),
        }

        const headers: HeadersInit = {}
        if (config.token) headers['Authorization'] = `Bearer ${config.token}`

        void fetch(config.url, {
          method: 'POST',
          headers,
          body: JSON.stringify(body),
        })
      }),
    )
  }

  return {
    dispatcher,
  }
}
