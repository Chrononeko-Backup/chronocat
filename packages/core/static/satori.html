<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <title>恭喜 - Chronocat</title>
    <style type="text/css">
      * {
        /* Font */
        font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI',
          'Microsoft YaHei UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans',
          'Noto Sans', 'Helvetica Neue', Helvetica, sans-serif;
        font-weight: 400;

        /* Font & text reset */
        font-style: normal;
        font-variant-ligatures: normal;
        font-variant-caps: normal;
        font-variant-numeric: normal;
        font-variant-east-asian: normal;
        font-stretch: normal;
        text-rendering: optimizeLegibility;
        text-indent: 0;
        text-shadow: none;
        text-decoration: none;
        writing-mode: horizontal-tb;
        letter-spacing: normal;
        word-spacing: normal;

        /* Layout */
        margin: 0;
        padding: 0;

        /* Box */
        box-sizing: border-box;
        border: none;

        /* Color */
        background-color: transparent;

        /* Transitions */
        transition: all 0.3s ease;
      }

      :root,
      body,
      #root {
        height: 100%;
        width: 100%;
      }

      :root {
        background-color: #1f1f1f;
        color: white;
      }

      body {
        display: flex;
        flex-direction: column;
        align-items: center;
        justify-content: center;
      }

      #root {
        max-width: 60vw;
        padding: 24px;

        display: flex;
        flex-direction: column;
        align-items: center;
        justify-content: center;
        row-gap: 24px;
      }

      h1 {
        font-weight: bold;
      }

      a {
        border-radius: 36px;
        padding: 20px 36px;

        background: #0d3b5e;

        font-size: 18px;
        font-weight: bold;

        color: white;
      }

      a:hover {
        background: #104b79;
      }

      a:active {
        background: #0d3b5e;
      }
    </style>
  </head>
  <body>
    <div id="root">
      <h1>恭喜！</h1>
      <p>
        你已经成功运行了 Chronocat。接下来，你可以参考 Chronocat 文档，将
        Chronocat 接入你喜爱的应用或框架。
      </p>
      <a href="https://chronocat.vercel.app/connect" rel="noopener noreferrer">
        前往接入
      </a>
    </div>
  </body>
</html>
