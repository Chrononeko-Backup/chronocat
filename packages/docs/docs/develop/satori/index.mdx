---
title: 开始
---

import Link from '@docusaurus/Link'
import { AreaDevelop } from '@site/src/components/AreaDevelop'

<AreaDevelop />

欢迎来到 Chronocat 的 Satori Protocol 协议文档！Satori 是 Chronocat
推荐使用的协议，基于 Satori 开发可以带来最佳的开发体验。

在本站以外，你还可以前往 [Satori 官网](https://satori.js.org/zh-CN/) 了解和学习 Satori。

## 快速起步

如果你想快速接入 Satori 并实现收发消息，请转到「快速起步」。

<Link
  className="button button--secondary button--lg"
  to="/develop/satori/quickstart"
>
  <span>快速起步</span>
</Link>

## 基础

如果你想从 Satori 的基础概念开始了解，请转到「基础」。

<Link
  className="button button--secondary button--lg"
  to="/develop/satori/fundamentals"
>
  <span>了解基础</span>
</Link>
