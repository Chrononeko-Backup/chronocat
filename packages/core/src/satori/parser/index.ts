import Element from '@satorijs/element'
import { Buffer } from 'node:buffer'
import type { ChronocatSatoriEventsConfig } from '../../config/types'
import type { Message as RedMessage } from '../../red'
import { ChatType, MsgType, SendType } from '../../red'
import type { Channel, Event, Guild, GuildMember } from '../types'
import { ChannelType } from '../types'
import { parseMsgTypes } from './msgt'

export const buildParser =
  (selfId: string, config: ChronocatSatoriEventsConfig) =>
  (message: RedMessage) =>
    parseMessageRecv(selfId, config, message)

export const parseMessageRecv = async (
  self_id: string,
  config: ChronocatSatoriEventsConfig,
  message: RedMessage,
) => {
  const parsed = await parseMessage(self_id, config, message)

  if (!parsed) return undefined

  const result: Event[] = []

  for (const event of parsed) {
    if (!event.message?.id && !event.user?.id) {
      result.push({
        id: undefined as unknown as number,
        platform: config.platform!,
        self_id,
        timestamp: Number(message.msgTime) * 1000,

        type: 'chrono-unsafe-warning-2127',
      })
      continue
    } else if (!event.message?.id)
      result.push({
        id: undefined as unknown as number,
        platform: config.platform!,
        self_id,
        timestamp: Number(message.msgTime) * 1000,

        type: 'chrono-unsafe-warning-2128',
        user: {
          id: event.user?.id as string,
          name: event.user?.name as string,
        },
      })
    else if (!event.user?.id)
      result.push({
        id: undefined as unknown as number,
        platform: config.platform!,
        self_id,
        timestamp: Number(message.msgTime) * 1000,

        type: 'chrono-unsafe-warning-2129',
        user: {
          id: undefined as unknown as string,
          name: event.user?.name as string,
        },
        message: {
          id: event.message?.id,
          content: undefined as unknown as string,
        },
      })
    else if (!event.user?.name && !event.member?.name)
      result.push({
        id: undefined as unknown as number,
        platform: config.platform!,
        self_id,
        timestamp: Number(message.msgTime) * 1000,

        type: 'chrono-unsafe-warning-2130',
        user: {
          id: event.user?.id,
          name: undefined as unknown as string,
        },
        message: {
          id: event.message?.id,
          content: undefined as unknown as string,
        },
      })

    result.push(event)
  }

  return result
}

export const parseMessage = async (
  self_id: string,
  config: ChronocatSatoriEventsConfig,
  message: RedMessage,
) => {
  const event: Event = {
    id: undefined as unknown as number,
    type: undefined as unknown as string,

    platform: config.platform!,
    self_id,
    timestamp: Number(message.msgTime) * 1000,
  }

  event.user = {
    id: message.senderUin,
    name: (message.sendNickName || undefined) as unknown as string,
    avatar: `http://q.qlogo.cn/headimg_dl?dst_uin=${message.senderUin}&spec=640`,
  }

  const ntMsgTypes = parseMsgTypes(message)

  // 无论哪种消息都有 Channel 和 User
  event.channel = {} as Channel

  // 判断消息来源
  switch (ntMsgTypes.chatType) {
    case ChatType.Private:
      event.channel.type = ChannelType.DIRECT
      event.channel.id = `private:${event.user.id}`
      event.channel.name = event.user.name
      break

    case ChatType.Group:
      // Guild 和 Member 只有群聊有
      event.guild = {} as Guild
      event.member = {} as GuildMember

      if (message.sendMemberName) event.member.name = message.sendMemberName

      event.channel.type = ChannelType.TEXT
      event.channel.id = event.guild.id = message.peerUid
      event.channel.name = event.guild.name = message.peerName
      event.guild.avatar = `https://p.qlogo.cn/gh/${message.peerUid}/${message.peerUid}/640`
      break
  }

  if (
    ntMsgTypes.msgType === MsgType.Normal &&
    ntMsgTypes.subMsgType.face &&
    message.elements.some((x) => x?.faceElement?.faceType === 5)
  )
    // 窗口抖动 / 手机按钮戳一戳
    // TODO
    return undefined
  else if (
    ntMsgTypes.msgType === MsgType.Ark &&
    message.subMsgType === 0 &&
    ntMsgTypes.sendType === SendType.Normal
  )
    // ARK 卡片消息，elementType = 10
    // 不处理
    return undefined
  else if (
    ntMsgTypes.msgType === MsgType.Normal ||
    ntMsgTypes.msgType === MsgType.Value3 ||
    ntMsgTypes.msgType === MsgType.Ptt ||
    ntMsgTypes.msgType === MsgType.Video ||
    ntMsgTypes.msgType === MsgType.WithRecords ||
    ntMsgTypes.msgType === MsgType.Vaule17
  )
    return parseChatMessage(self_id, config, event, message).then((x) => [
      x[0],
      ...x[1],
    ])
  // else if (event.__CHRONO_UNSAFE_NTMSGTYPES__.subMsgType.multiForward)
  //   // 合并转发消息
  //   // multiForwardMsgElement（elementType = 16）内并不带有合并转发的全部内容，
  //   // 需要后续通过 API 再请求
  //   // 考虑到合并转发消息解析需求较少，不调度此 session
  //   return undefined
  else if (
    ntMsgTypes.msgType === MsgType.System && // 5
    message.subMsgType === 8 && // 8
    ntMsgTypes.sendType === SendType.System && // 3
    message.elements[0]!.elementType === 8 && // 8
    message.elements[0]!.grayTipElement!.subElementType === 4 && // 4
    message.elements[0]!.grayTipElement!.groupElement!.type === 1 // 1
  )
    // 新人自行入群
    return await parseGuildMemberAddedMessage(self_id, config, event, message)
  else if (
    ntMsgTypes.msgType === MsgType.System && // 5
    message.subMsgType === 8 && // 8
    ntMsgTypes.sendType === SendType.System && // 3
    message.elements[0]!.elementType === 8 && // 8
    message.elements[0]!.grayTipElement!.subElementType === 4 && // 4
    message.elements[0]!.grayTipElement!.groupElement!.type === 8 // 8
  )
    // 他人被禁言
    return await parseGuildMemberMuteMessage(self_id, config, event, message)
  else if (
    ntMsgTypes.msgType === MsgType.System && // 5
    message.subMsgType === 8 && // 8
    ntMsgTypes.sendType === SendType.System && // 3
    message.elements[0]!.elementType === 8 && // 8
    message.elements[0]!.grayTipElement!.subElementType === 4 && // 4
    message.elements[0]!.grayTipElement!.groupElement!.type === 5 // 5
  )
    // 群名称变更
    return undefined
  else if (
    ntMsgTypes.msgType === MsgType.System && // 5
    message.subMsgType === 12 && // 12
    ntMsgTypes.sendType === SendType.System && // 3
    message.elements[0]!.elementType === 8 && // 8
    message.elements[0]!.grayTipElement!.subElementType === 12 && // 12
    message.elements[0]!.grayTipElement!.xmlElement!.busiType === '1' && // 1
    message.elements[0]!.grayTipElement!.xmlElement!.busiId === '10145' // 10145
  )
    // 旧版群成员邀请新人入群
    return await parseGuildMemberAddedLegacyInviteMessage(
      self_id,
      config,
      event,
      message,
    )
  else if (
    ntMsgTypes.msgType === MsgType.System &&
    message.subMsgType === 17 &&
    ntMsgTypes.sendType === SendType.System
  )
    // 群主禁止群内临时通话
    // 群主禁止群内发起新的群聊
    return undefined

  return undefined
}

/**
 * 解析聊天消息。
 *
 * @remarks
 * 在消息没有除了消息元素以外的其他属性需要处理的情况下，直接使用此方法。
 */
async function parseChatMessage(
  self_id: string,
  config: ChronocatSatoriEventsConfig,
  event: Event,
  message: RedMessage,
) {
  const [elements, extraEvents] = await parseElements(self_id, config, message)
  event.type = 'message-created'
  event.message = {
    id: message.msgId,
    content: elements.join(''),
  }
  return [event, extraEvents] as const
}

/**
 * 解析新人自行入群消息。
 *
 * @remarks
 * 通过解析 `grayTipElement`（elementType = 8）中的
 * `groupElement`（subElementType = 4）即可直接提取 QQ 号。
 */
async function parseGuildMemberAddedMessage(
  self_id: string,
  config: ChronocatSatoriEventsConfig,
  event: Event,
  message: RedMessage,
) {
  const [event2, extraEvents] = await parseChatMessage(
    self_id,
    config,
    event,
    message,
  )
  event2.type = 'guild-member-added'

  event2.operator = {
    id: message.elements[0]!.grayTipElement!.groupElement!.adminUin!,
    name: undefined as unknown as string,
  }

  event2.user = {
    id: message.elements[0]!.grayTipElement!.groupElement!.memberUin!,
    name: message.elements[0]!.grayTipElement!.groupElement!.memberNick!,
    avatar: `http://q.qlogo.cn/headimg_dl?dst_uin=${
      message.elements[0]!.grayTipElement!.groupElement!.memberUin
    }&spec=640`,
  }

  if (event2.member) delete event2.member

  return [event2, ...extraEvents]
}

/**
 * 解析他人被禁言消息。
 */
async function parseGuildMemberMuteMessage(
  self_id: string,
  config: ChronocatSatoriEventsConfig,
  event: Event,
  message: RedMessage,
) {
  const [event2, extraEvents] = await parseChatMessage(
    self_id,
    config,
    event,
    message,
  )
  if (
    Number(message.elements[0]!.grayTipElement!.groupElement!.shutUp!.duration)
  )
    event2.type = 'chrono-unsafe-guild-mute'
  else event2.type = 'chrono-unsafe-guild-unmute'

  event2.operator = {
    id: message.elements[0]!.grayTipElement!.groupElement!.shutUp!.admin.uin,
    name: undefined as unknown as string,
  }

  event2.user = {
    id: message.elements[0]!.grayTipElement!.groupElement!.shutUp!.member.uin,
    name: message.elements[0]!.grayTipElement!.groupElement!.shutUp!.member
      .name,
    avatar: `http://q.qlogo.cn/headimg_dl?dst_uin=${
      message.elements[0]!.grayTipElement!.groupElement!.shutUp!.member.uin
    }&spec=640`,
  }

  if (event2.member) delete event2.member

  return [event2, ...extraEvents]
}

const regexGuildMemberAddedLegacyInviteMessage = /jp="(\d+)".*jp="(\d+)"/gim

/**
 * 解析旧版群成员邀请新人入群消息。使用 NT
 * 以前的客户端邀请他人加群会收到此消息。
 *
 * @remarks
 * 遗憾地，旧版群成员邀请新人入群消息不存在能够直接获取成员 QQ
 * 号的方法。需要通过解析 `grayTipElement`（elementType = 8）中的
 * `xmlElement`（subElementType = 12）中的 HTML 来提取 QQ 号。
 *
 * 一个 HTML 的示例：
 *
 * ```html
 * <gtip align="center">
 *   <qq uin="u_0gvBEjIEEOk5-EypJRjwxw" col="3" jp="1302744182" />
 *   <nor txt="邀请"/>
 *   <qq uin="u_ENIKFfFS74WSiKNoA6ERWg" col="3" jp="2953529126" />
 *   <nor txt="加入了群聊。"/>
 * </gtip>
 * ```
 *
 * 目前使用正则进行 QQ 号的提取。如果未来正则失效，则应考虑使用 cheerio
 * 进行提取。
 */
async function parseGuildMemberAddedLegacyInviteMessage(
  self_id: string,
  config: ChronocatSatoriEventsConfig,
  event: Event,
  message: RedMessage,
) {
  const [event2, extraEvents] = await parseChatMessage(
    self_id,
    config,
    event,
    message,
  )
  event2.type = 'guild-member-added'

  const execArr = regexGuildMemberAddedLegacyInviteMessage.exec(
    message.elements[0]!.grayTipElement!.xmlElement!.content!,
  )
  if (!Array.isArray(execArr) || execArr.length < 3) return undefined
  const [_, operatorId, userId] = execArr

  event2.operator = {
    id: operatorId,
    name: undefined as unknown as string,
  }

  event2.user = {
    id: userId,
    name: undefined as unknown as string,
    avatar: `http://q.qlogo.cn/headimg_dl?dst_uin=${userId}&spec=640`,
  }

  if (event2.member) delete event2.member

  return [event2, ...extraEvents]
}

/**
 * 解析消息元素。
 */
async function parseElements(
  self_id: string,
  config: ChronocatSatoriEventsConfig,
  message: RedMessage,
) {
  const elements: Element[] = []
  const extraEvents: Event[] = []

  for (const m of message.elements) {
    switch (m.elementType) {
      case 1: {
        // 文本消息
        switch (m.textElement!.atType) {
          case 0: {
            // 纯文本消息
            elements.push(Element.text(m.textElement?.content))
            break
          }

          case 2: {
            // at 消息
            const id = m.textElement!.atNtUin
            const name = m.textElement!.content.slice(1)
            if (!id) {
              extraEvents.push({
                id: undefined as unknown as number,
                platform: config.platform!,
                self_id,
                timestamp: Number(message.msgTime) * 1000,

                type: 'chrono-unsafe-warning-2131',
                user: {
                  id: undefined as unknown as string,
                  name,
                },
              })
              break
            }
            elements.push(
              Element('at', {
                id,
                name,
              }),
            )
            break
          }
        }
        break
      }

      case 2: {
        // 图片消息
        elements.push(
          Element('img', {
            src: `${config.self_url}/v1/assets/${Buffer.from(
              JSON.stringify({
                msgId: message.msgId,
                chatType: message.chatType,
                peerUid: message.peerUid,
                elementId: m.elementId,
                thumbSize: m.picElement!.thumbFileSize,
              }),
            ).toString('base64url')}`,
            'chrono-unsafe-isemoji':
              m.picElement!.picSubType === 1 || undefined,
            'chrono-unsafe-filepath': m.picElement!.sourcePath,
            'chrono-unsafe-filetype': m.picElement!.picType,
          }),
        )
        break
      }

      case 3: {
        // 文件消息
        elements.push(
          Element('file', {
            src: `${config.self_url}/v1/assets/${Buffer.from(
              JSON.stringify({
                msgId: message.msgId,
                chatType: message.chatType,
                peerUid: message.peerUid,
                elementId: m.elementId,
                thumbSize: m.fileElement!.thumbFileSize,
              }),
            ).toString('base64url')}`,
            'chrono-unsafe-filename': m.fileElement!.fileName,
          }),
        )
        break
      }

      case 4: {
        // 语音消息
        elements.push(
          Element('audio', {
            src: `${config.self_url}/v1/assets/${Buffer.from(
              JSON.stringify({
                msgId: message.msgId,
                chatType: message.chatType,
                peerUid: message.peerUid,
                elementId: m.elementId,
                thumbSize: 0,
              }),
            ).toString('base64url')}`,
            'chrono-unsafe-filepath': m.pttElement!.filePath,
            'chrono-unsafe-filename': m.pttElement!.fileName,
            'chrono-unsafe-time': m.pttElement!.duration,
            'chrono-unsafe-voice-change-type': m.pttElement!.voiceChangeType,
            'chrono-unsafe-wave-amplitudes':
              m.pttElement!.waveAmplitudes.join(),
          }),
        )
        break
      }

      case 5: {
        // 视频消息
        elements.push(
          Element('video', {
            src: `${config.self_url}/v1/assets/${Buffer.from(
              JSON.stringify({
                msgId: message.msgId,
                chatType: message.chatType,
                peerUid: message.peerUid,
                elementId: m.elementId,
                thumbSize: m.videoElement!.thumbSize,
              }),
            ).toString('base64url')}`,
            'chrono-unsafe-filepath': m.videoElement!.filePath,
            'chrono-unsafe-filename': m.videoElement!.fileName,
            'chrono-unsafe-file-format': m.videoElement!.fileFormat,
            'chrono-unsafe-time': m.videoElement!.fileTime,
          }),
        )
        break
      }

      default:
        break
    }
  }

  return [elements, extraEvents] as const
}
