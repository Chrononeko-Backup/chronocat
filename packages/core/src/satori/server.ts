import type { IncomingMessage, ServerResponse } from 'node:http'
import { createServer } from 'node:http'
import type { WebSocket } from 'ws'
import { WebSocketServer } from 'ws'
import index from '../../static/satori.html'
import type { ChronocatSatoriServerConfig } from '../config/types'
import type { DispatchMessage } from '../dispatch'
import { selfProfile } from '../ipc/globalVars'
import { getAuthData } from '../utils/authData'
import { timeout } from '../utils/time'
import { buildEventIdCounter } from '../utils/token'
import type { Routes } from './routes'
import { routes } from './routes'
import { assets } from './routes/assets'
import type { RouteContext } from './routes/types'
import type { Login, WebSocketIncomingMessage } from './types'
import { LoginStatus, Op } from './types'

declare const __DEFINE_CHRONO_VERSION__: string

const prefix = '/v1/'
const poweredBy = `Chronocat/${__DEFINE_CHRONO_VERSION__}`

export const initSatoriServer = async (config: ChronocatSatoriServerConfig) => {
  // 预处理 self_url
  if (!config.self_url || config.self_url === 'https://chronocat.vercel.app')
    config.self_url = `http://127.0.0.1:${config.port}`
  if (config.self_url.endsWith('/'))
    config.self_url = config.self_url.slice(0, config.self_url.length - 1)

  const authorizedClients: WebSocket[] = []

  const getId = buildEventIdCounter()

  // eslint-disable-next-line @typescript-eslint/no-misused-promises
  const server = createServer(async (req, res) => {
    if (!req.url) {
      res.writeHead(400)
      res.end('404 bad request')
      return
    }

    const url = new URL(req.url, `http://${req.headers.host}`)

    res.setHeader('Server', poweredBy)
    res.setHeader('X-Powered-By', poweredBy)

    if (url.pathname === '/') {
      res.writeHead(200, {
        'Content-Type': 'text/html; charset=UTF-8',
        'Cache-Control': 'no-cache',
        'Content-Length': index.byteLength,
      })
      res.end(index)
    }

    if (!url.pathname.startsWith(prefix)) {
      res.writeHead(404)
      res.end('404 not found')
      return
    }

    const path = url.pathname.slice(prefix.length)

    if (path.startsWith('assets/')) {
      if (req.method !== 'GET') {
        res.writeHead(405)
        res.end('405 method not allowed')
        return
      }

      try {
        await assets({
          raw: path.slice('assets/'.length),
          ...buildRouteCtx(config, req, res),
        })

        if (!res.writableEnded) throw new Error()

        return
      } catch (e) {
        console.log(e)

        res.writeHead(500)
        res.end('500 internal server error')

        return
      }
    }

    if (
      config.token &&
      !(
        req.headers.authorization?.slice(0, 7) === 'Bearer ' &&
        req.headers.authorization.slice(7) === config.token
      )
    ) {
      res.writeHead(401)
      res.end('401 unauthorized')
      return
    }

    const method = routes[path as Routes]

    if (!method) {
      res.writeHead(404)
      res.end('404 not found')
      return
    }

    if (req.method !== 'POST') {
      res.writeHead(405)
      res.end('405 method not allowed')
      return
    }

    try {
      const result = await method(buildRouteCtx(config, req, res))

      if (!res.writableEnded) {
        res.writeHead(200, {
          'Content-Type': 'application/json',
        })
        res.end(JSON.stringify(result))
      }

      return
    } catch (e) {
      console.log(e)

      res.writeHead(500)
      res.end('500 internal server error')

      return
    }
  })

  const wsServer = new WebSocketServer({
    server,
    path: prefix + 'events',
  })

  // eslint-disable-next-line @typescript-eslint/no-misused-promises
  wsServer.on('connection', async (ws) => {
    let authorized = !config.token

    const addToAuthorizedClients = () => {
      authorizedClients.push(ws)
      ws.on('close', () =>
        authorizedClients.splice(authorizedClients.indexOf(ws), 1),
      )
    }

    // eslint-disable-next-line @typescript-eslint/no-misused-promises
    ws.on('message', async (raw) => {
      const message = JSON.parse(
        // eslint-disable-next-line @typescript-eslint/no-base-to-string
        raw.toString(),
      ) as WebSocketIncomingMessage

      switch (message.op) {
        case Op.Ping: {
          ws.send(
            JSON.stringify({
              op: Op.Pong,
            }),
          )
          return
        }

        case Op.Identify: {
          if (authorized) {
            ws.send(
              JSON.stringify({
                op: Op.Event,
                body: {
                  id: getId(),
                  platform: config.platform!,
                  self_id: selfProfile.value!.uin,
                  timestamp: new Date().getTime(),
                  type: 'chrono-unsafe-warning-2132',
                },
              }),
            )
            return
          }

          if (config.token && message.body?.token !== config.token) {
            ws.close(3000, 'Unauthorized')
            return
          }

          authorized = true
          addToAuthorizedClients()

          const login: Login = {
            user: {
              id: selfProfile.value!.uin,
              name: selfProfile.value!.nick,
              is_bot: true,
            },
            self_id: selfProfile.value!.uin,
            platform: config.platform!,
            status: LoginStatus.ONLINE,
          }

          ws.send(
            JSON.stringify({
              op: Op.Ready,
              body: {
                logins: [login],
              },
            }),
          )

          return
        }

        default: {
          ws.send(
            JSON.stringify({
              op: Op.Event,
              body: {
                id: getId(),
                platform: config.platform!,
                self_id: selfProfile.value!.uin,
                timestamp: new Date().getTime(),
                type: 'chrono-unsafe-warning-2133',
              },
            }),
          )
          return
        }
      }
    })

    if (config.token)
      setTimeout(() => {
        if (!authorized) ws.close(3000, 'Unauthorized')
      }, timeout)
    else addToAuthorizedClients()
  })

  const dispatcher = (message: DispatchMessage) =>
    // eslint-disable-next-line @typescript-eslint/no-misused-promises
    authorizedClients.forEach(async (ws) => {
      await message.toSatori((await getAuthData()).uin, config).then((events) =>
        events.forEach((body) =>
          ws.send(
            JSON.stringify({
              op: Op.Event,
              body: {
                ...body,
                id: getId(),
              },
            }),
          ),
        ),
      )
    })

  server.listen(config.port, config.listen)

  return {
    dispatcher,
  }
}

function buildRouteCtx(
  config: ChronocatSatoriServerConfig,
  req: IncomingMessage,
  res: ServerResponse<IncomingMessage>,
): RouteContext {
  const buffer = () => {
    const chunks: Buffer[] = []
    return new Promise<Buffer>((resolve, reject) => {
      req.on('data', (chunk) => {
        chunks.push(chunk as Buffer)
      })
      req.on('end', () => {
        resolve(Buffer.concat(chunks))
      })
      req.on('error', () => {
        reject()
      })
    })
  }

  const string = () => buffer().then((b) => b.toString('utf-8'))

  const json = () => string().then((s) => JSON.parse(s) as unknown)

  return {
    config,
    req,
    res,
    buffer,
    string,
    json,
  }
}
