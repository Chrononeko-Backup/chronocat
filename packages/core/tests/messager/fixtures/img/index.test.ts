import { resolve } from 'node:path'
import { pathToFileURL } from 'node:url'
import { Messager } from '../../../../src/satori/messager'
import { getMockMessage, satoriConfig, saveResult } from '../../../mocks'

test('Satori编码器应当正确编码 图片消息', async () => {
  const send = jest.fn().mockReturnValueOnce(getMockMessage())
  const save = jest.fn().mockReturnValueOnce(saveResult)

  const sendResult = await new Messager(
    '9999',
    satoriConfig,
    {
      send,
      save,
    },
    '9998',
  ).send(
    `xxx<img src="${pathToFileURL(
      resolve(__dirname, '../../../../docs/static/chronocat.png'),
    ).toString()}" />yyy`,
  )

  expect(send.mock.calls).toMatchSnapshot()

  expect(save.mock.calls).toHaveLength(1)
  expect(save.mock.calls[0]).toHaveLength(1)
  expect((save.mock.calls[0] as [string])[0]).toMatch(
    /\/docs\/static\/chronocat.png$/,
  )

  expect(sendResult).toHaveLength(1)
  expect(sendResult[0].id).toBe('1000000000000000000')
})
