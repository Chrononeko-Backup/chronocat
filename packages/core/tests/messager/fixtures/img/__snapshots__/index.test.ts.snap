// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Satori编码器应当正确编码 图片消息 1`] = `
[
  [
    {
      "chatType": 2,
      "guildId": "",
      "peerUin": "9998",
    },
    [
      {
        "elementId": "",
        "elementType": 1,
        "textElement": {
          "atNtUid": "",
          "atTinyId": "",
          "atType": 0,
          "atUid": "",
          "content": "xxx",
        },
      },
      {
        "elementId": "",
        "elementType": 1,
        "textElement": {
          "atNtUid": "",
          "atTinyId": "",
          "atType": 0,
          "atUid": "",
          "content": "yyy",
        },
      },
    ],
  ],
]
`;
